docs/content/sips/000-spin-test.md

```diff
@@ -0,0 +1,190 @@
+title = "SIP 000 - `spin-test`"
+template = "main"
+date = "2024-04-18"
+---
+
+Summary: create a component based testing framework for testing Spin applications within a WebAssembly sandbox.
+
+Owner(s): ryan.levick@fermyon.com
+
+Created: April 18, 2024
+
+This SIP describes the `spin-test` tool that allows running unmodified Spin applications in WebAssembly component based testing environment where all potential imports to the Spin app are virtualized *and* can be configured by the user in a language independent way.
+
+Implementation of this SIP has already begun [here](https://github.com/fermyon/spin-test).
+
+## Requirements
+
+The requirements for such a tool are the following:
+
+- Does not require the user to modify their application in any way - what runs in the test is what runs in production.
+- The app and all the imports to the app run in WebAssembly and not on the host.
+- The test itself also runs in WebAssembly allowing tests to be written in any language with WebAssembly component guest support.
+- Allows the user to modify the testing environment in a programmatic way in any language they want to use.
+- Have the test environment match functionality as much as is possible with other Spin runtimes such as found in Spin CLI, Fermyon Cloud, and SpinKube.
+
+## Proposal
+
+To achieve the goals above, we propose a new tool tentatively called `spin-test`.
+
+The `spin-test` binary would be callable from any Spin project in much the same way that `spin up` is project aware (i.e., it doesn’t need to be told where the `spin.toml` manifest is). 
+
+Invoking `spin-test` will automatically resolve where the Spin app component lives. The tool will automatically find tests that should be run and will run those tests against the Spin app component reporting whether the tests passed or failed. `spin-test` will come with basic ways of filtering tests so that a subset of tests can be run.
+
+**Note:** See the *Open Questions* section for discussion on how `spin-test` discovers tests to run.
+
+### Implementation
+
+At the core of `spin-test` is component composition. The user’s Spin app is composed together with other components to ultimately produce a component (referred to as “the composition” below) that has none of the original Spin or WASI imports. The composition has the following components internal to it:
+
+- A single component from the Spin app (see the *Open Questions* section for why only one component from a Spin app can be supported)
```

Comment:
This refers to "a single component" from the perspective of the Spin application manifest I assume — if my Spin component build process generates a statically linked component, I would expect testing to continue to work as expected.

Is that accurate?